service/controller/resource/tcnpsecuritygroups/resource.go

```diff
@@ -0,0 +1,199 @@
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	r := &Resource{
+		logger: config.Logger,
+	}
+
+	return r, nil
+}
+
+func (r *Resource) Name() string {
+	return Name
+}
+
+func (r *Resource) addInfoToCtx(ctx context.Context, cr infrastructurev1alpha2.AWSMachineDeployment) error {
+	cc, err := controllercontext.FromContext(ctx)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	var desiredSecurityGroupIDs []string
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding all np security desiredSecurityGroupIDs for tenant cluster %#q", key.ClusterID(&cr)))
+
+		i := &ec2.DescribeSecurityGroupsInput{
+			Filters: []*ec2.Filter{
+				{
+					Name: aws.String(fmt.Sprintf("tag:%s", key.TagCluster)),
+					Values: []*string{
+						aws.String(key.ClusterID(&cr)),
+					},
+				},
+				{
+					Name: aws.String(fmt.Sprintf("tag:%s", key.TagStack)),
+					Values: []*string{
+						aws.String(key.StackTCNP),
+					},
+				},
+			},
+		}
+		o, err := cc.Client.TenantCluster.AWS.EC2.DescribeSecurityGroups(i)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		// check if the node pools is not being deleted
+		// we check for ec2 instances for that node pool machine deployment
+		for _, sg := range o.SecurityGroups {
+
+			var machineDeploymentID string
+			{
+				for _, tag := range sg.Tags {
+					if *tag.Key == key.TagMachineDeployment {
+						machineDeploymentID = *tag.Value
+						break
+					}
+				}
+			}
+			// ignore security group for this node pool machine deployment
+			if machineDeploymentID == key.MachineDeploymentID(&cr) {
+				continue
+			}
+
+			i := &ec2.DescribeInstancesInput{
+				Filters: []*ec2.Filter{
+					{
+						Name: aws.String(fmt.Sprintf("tag:%s", key.TagMachineDeployment)),
+						Values: []*string{
+							aws.String(machineDeploymentID),
+						},
+					},
+					{
+						Name: aws.String("instance-state-name"),
+						Values: []*string{
+							aws.String(ec2.InstanceStateNamePending),
+							aws.String(ec2.InstanceStateNameRunning),
+							aws.String(ec2.InstanceStateNameStopped),
+							aws.String(ec2.InstanceStateNameStopping),
+						},
+					},
+				},
+			}
+
+			o, err := cc.Client.TenantCluster.AWS.EC2.DescribeInstances(i)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+			// add security group to the list only if the node pool machine deployment has any running instances
+			// if there are no running or pending instances the node pool machine deployment might be deleted
+			// and we want to remove the sg rules as well
+
+			if len(o.Reservations) > 0 && len(o.Reservations[0].Instances) > 0 {
+				desiredSecurityGroupIDs = append(desiredSecurityGroupIDs, *sg.GroupId)
+			}
+		}
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found %d np security desiredSecurityGroupIDs for machine deployment%#q", len(desiredSecurityGroupIDs), key.MachineDeploymentID(&cr)))
```

Comment:
```suggestion

		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found %d desired node pool security groups for cluster %#q", len(desiredSecurityGroupIDs), key.ClusterID(&cr)))
```

Reply:
feel like there should be machine deployment ID as for each cluster this message will print several times depending on machine deployment count

Reply:
The machine deployment ID is already in the object link of the logs. The first logs also started with the cluster ID and then mixed it. I tried to make it consistent and what we actually miss in the logs is the cluster ID.